pilot/pkg/security/authz/builder/builder.go

```diff
@@ -0,0 +1,161 @@
+	authzPolicies               *model.AuthorizationPolicies
+	isGlobalPermissiveEnabled   bool // True if global RBAC config is in permissive mode.
+	isXDSMarshalingToAnyEnabled bool
+}
+
+func NewBuilder(serviceInstance *model.ServiceInstance, policies *model.AuthorizationPolicies, isXDSMarshalingToAnyEnabled bool) *Builder {
+	if serviceInstance.Service == nil {
+		rbacLog.Errorf("no service for serviceInstance: %v", serviceInstance)
+		return nil
+	}
+
+	serviceName := serviceInstance.Service.Attributes.Name
+	serviceNamespace := serviceInstance.Service.Attributes.Namespace
+	serviceHostname := string(serviceInstance.Service.Hostname)
+	if !isRbacEnabled(serviceHostname, serviceNamespace, policies) {
+		rbacLog.Debugf("RBAC disabled for service %s", serviceHostname)
+		return nil
+	}
+
+	serviceMetadata, err := authz_model.NewServiceMetadata(serviceName, serviceNamespace, serviceInstance)
+	if err != nil {
+		rbacLog.Errorf("failed to create ServiceMetadata for %s: %s", serviceName, err)
+		return nil
+	}
+
+	rbacConfig := policies.RbacConfig
+	isGlobalPermissiveEnabled := rbacConfig != nil && rbacConfig.EnforcementMode == istio_rbac.EnforcementMode_PERMISSIVE
+
+	return &Builder{
+		serviceMetadata:             serviceMetadata,
+		authzPolicies:               policies,
+		isGlobalPermissiveEnabled:   isGlobalPermissiveEnabled,
+		isXDSMarshalingToAnyEnabled: isXDSMarshalingToAnyEnabled,
+	}
+}
+
+func (b *Builder) BuildHTTPFilter() *http_filter.HttpFilter {
+	rbacConfig := b.build(false /* forTCPFilter */)
+	httpConfig := http_filter.HttpFilter{
+		Name: authz_model.RBACHTTPFilterName,
+	}
+	if b.isXDSMarshalingToAnyEnabled {
+		httpConfig.ConfigType = &http_filter.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(rbacConfig)}
+	} else {
+		httpConfig.ConfigType = &http_filter.HttpFilter_Config{Config: util.MessageToStruct(rbacConfig)}
+	}
+
+	rbacLog.Debugf("built http filter config: %v", httpConfig)
+	return &httpConfig
+}
+
+func (b *Builder) BuildTCPFilter() *tcp_filter.Filter {
+	// The build function always return the config for HTTP filter, we need to extract the
+	// generated rules and set it in the config for TCP filter.
+	config := b.build(true /* forTCPFilter */)
+	rbacConfig := &tcp_config.RBAC{
+		Rules:       config.Rules,
+		ShadowRules: config.ShadowRules,
+		StatPrefix:  authz_model.RBACTCPFilterStatPrefix,
+	}
+
+	tcpConfig := tcp_filter.Filter{
+		Name: authz_model.RBACTCPFilterName,
+	}
+	if b.isXDSMarshalingToAnyEnabled {
+		tcpConfig.ConfigType = &tcp_filter.Filter_TypedConfig{TypedConfig: util.MessageToAny(rbacConfig)}
+	} else {
+		tcpConfig.ConfigType = &tcp_filter.Filter_Config{Config: util.MessageToStruct(rbacConfig)}
+	}
+
+	rbacLog.Debugf("built tcp filter config: %v", tcpConfig)
+	return &tcpConfig
+}
+
+// isServiceInList checks if a given service or namespace is found in the RbacConfig target.
+func isServiceInList(serviceHostname string, namespace string, target *istio_rbac.RbacConfig_Target) bool {
```

Comment:
Maybe change `isServiceInList` to something like `isRbacEnforcedOnService` to make it more clear what it's doing, since this function is to check if RBAC is enforcing on this service.

Reply:
We already have a `isRbacEnabled` function, I'll just change this one to `isInRbacTargetList` for simplicity.